### ISSUE TITLE
Runtime error (IndexError ) linting file with jinja "if"

### ISSUE THREAD

**Issue by byapparov** [sqlfluff/sqlfluff#1414]
## Expected Behaviour

Get list of rule violations by line  for the file:

```sql
-- This file combines product data from individual brands into a staging table
{% set products =  [
  'tabe1',
  'table2'] %}

{% for product in products %}
SELECT
  brand,
  country_code,
  category,
  name,
  id
FROM
  {{ ref(product) }}
{% if not loop.last -%} UNION ALL {%- endif %}
{% endfor %}
```

## Observed Behaviour

Cannot make sqlfluff work with jinja for loops.

Error:

```
WARNING    Unable to lint /app/models/3_INFORMATION_MART/orders.sql due to an internal error. Please report this as an issue with your query's contents and stacktrace below!
To hide this warning, add the failing file to .sqlfluffignore
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/runner.py", line 94, in run
    yield partial()
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 512, in lint_rendered
    parsed = cls.parse_rendered(rendered)
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 294, in parse_rendered
    tokens, lvs, config = cls._lex_templated_file(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/linter/linter.py", line 127, in _lex_templated_file
    tokens, lex_vs = lexer.lex(templated_file)
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/parser/lexer.py", line 319, in lex
    segments: Tuple[RawSegment, ...] = self.elements_to_segments(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/parser/lexer.py", line 346, in elements_to_segments
    source_slice = templated_file.templated_slice_to_source_slice(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/templaters/base.py", line 287, in templated_slice_to_source_slice
    start_slices = [self.sliced_file[ts_start_sf_start]]
IndexError: list index out of range

```

## Steps to Reproduce

run lint against the target file

## Dialect
dbt

## Version
0.6.4

## Configuration
```
[sqlfluff]
templater = dbt
exclude_rules = L031,L029

[sqlfluff:rules]
tab_space_size = 2

[sqlfluff:templater:dbt]
profiles_dir = .
profile = dev

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = upper

```

**Comment by barrywhart** [comment]
I reproduced this error with the `jinja` templater, so it is not specific to dbt.

```
-- This file combines product data from individual brands into a staging table
{% set products =  [
  'tabe1',
  'table2'] %}

{% for product in products %}
SELECT
  brand,
  country_code,
  category,
  name,
  id
FROM
  product
{% if not loop.last -%} UNION ALL {%- endif %}
{% endfor %}
```

**Comment by barrywhart** [comment]
This line is triggering the error. The error goes away if I either:
* Add a newline prior to the `endif`
* Change `{%- endif %}` to `{% endif %}`

```
{% if not loop.last %} UNION ALL {%- endif %}
```

A recent PR (#1364) made some changes regarding how `{%-` is handled. I don't know that this bug is related, but it's at least _interesting_ that we've seen multiple bugs recently relating to Jinja "whitespace control".

**Comment by barrywhart** [comment]
This minimal example produces the same error:
```
{% set v = 0 %}

{% if not 1 %}{% endif %}
```
